### ISSUE TITLE
WAZUH *db-wal files in /var/ossec/queue/db not being cleared down [[Reopening of bug reported in  #3535 ]]

### ISSUE THREAD

**Issue by Andy13000ft** [wazuh/wazuh#4411]
Wazuh Version: 3.10.2
Component: Wazuh-db
Install Type: Manager
Install Method: Packages
Platform: Redhat Enterprise Linux 7.5
----------------------------------------
PROBLEM: - The *db-wal files in /var/ossec/queue/db are not being cleared down, even though commits are being made to the database. This results in space being used up and eventually client connections hanging. Simply restarting the manager process, which _does_ clear down the files, is not an acceptable solution 
----------------------------------------
NOTE: Please see original bug report #3535 that was closed. However a fix for this is still required as the suggested workaround is not suitable in all cases (see below)
----------------------------------------
Thanks for the information in this thread so far. I am working with Wazuh 3.10.2 and also have the problem of the WAL files from each of my agents growing uncontrollably. So far I have put the manager into debug mode and I can see the agents being committed, but the WAL file corresponding to that agent does not shrink in size. Many of my WAL files are above 4MB in size which I understood was the threshold for the file to be shrunk during a commit. I am also happy that there is not constant communication between the agent and the manager by observing the alerts log. Again there is a period of greater than 60 seconds when some of the agents, with large WAL files do not communicate, another trigger I understood to cause the files to shrink. However their files only shrink (back to zero) if the manager is restarted. Unfortunately we don't want to keep doing this as we lose alerts from agentless clients, i.e. routers etc. This lack of file shrinking seems to be a bug....any thoughts ?

**Comment by Theoooooo** [comment]
I was having the same issue with the 3.10.2 update for a while and suddenly it has stopped doing it.

Now after upgrading to 3.11.1 the issue is back and it has started filling my /var partition because of this

**Comment by Zenidd** [comment]
Hello @Theoooooo,

First of all sorry for the late reply. We are working really hard on different issues and development tasks.

> So far I have put the manager into debug mode and I can see the agents being committed, but the WAL file corresponding to that agent does not shrink in size.

We are investigating if another operation is needed after the commit, to ensure the DB is properly closed on any state.

> Again there is a period of greater than 60 seconds when some of the agents, with large WAL files do not communicate, another trigger I understood to cause the files to shrink. However their files only shrink (back to zero) if the manager is restarted. 

[We have already limited the commit interval](https://github.com/wazuh/wazuh/issues/3803) for upcoming versions so now it also has an upper limit. This fix should make the commits happen in its time.

Sorry for the inconvenience, we agree on the manager restarts not being a viable solution for WAL files purge. As a workaround, until we fix this bug, you should set the max number of opened databases to just 1 should force the DB to commit and the WAL files to shrink continuously.

`# echo "wazuh_db.open_db_limit=1" >> /var/ossec/etc/local_internal_options.conf`

Please, let me know if it helps. We will update this issue with the results of the investigations and with any data or clue we get.

Greetings, JP Sáez

**Comment by Andy13000ft** [comment]
JP,

Many thanks for the feedback. I will look to test the suggested alternate workaround asap and let you know.

Thanks again.
Regards,
Andy

**Comment by vikman90** [comment]
Hi @Andy13000ft,

We're taking a deeper look at this issue. We introduced a new option `wazuh_db.commit_time_max=60` to limit the time to commit a transaction in the databases. This works, but the WAL file does not shrink. That is because the database keeps up to 1000 pages in the WAL file:
>Of course, one wants to eventually transfer all the transactions that are appended in the WAL file back into the original database. Moving the WAL file transactions back into the database is called a “checkpoint”.
>(…)
>By default, SQLite does a checkpoint automatically when the WAL file reaches a threshold size of 1000 pages.

([SQLite 3 Write-Ahead Logging](https://www.sqlite.org/wal.html))

We can only modify this parameter in compile-time. However, WAL mode is necessary for the legacy global database, due to concurrent query support. We don't need it in the new database files (_/var/ossec/queue/db_) so we can use the default journaling mode (_REMOVE_):

> The DELETE journaling mode is the normal behavior. In the DELETE mode, the rollback journal is deleted at the conclusion of each transaction.

([Pragma journal_mode](https://www.sqlite.org/pragma.html#pragma_journal_mode))

I've checked that the journal file grows during a Syscollector report or an SCA scan:

```
-rw-r-----  1 ossec ossec 680K Apr 23 12:31 000.db
-rw-r-----  1 ossec ossec 318K Apr 23 12:31 000.db-journal
```

Then, the database daemon (Wazuh DB) commits the database and removes the journal file:

```
-rw-r-----  1 ossec ossec 956K Apr 23 12:31 000.db
```

## Proposed fix

- Use the default journaling file (_DELETE_) in Wazuh DB.
- Keep the _WAL_ journaling mode in the legacy databases.

We're working on that. Thank you very much for reporting this issue. We'll keep you posted.

Best regards.

**Comment by vikman90** [comment]
@Andy13000ft we've merged the fix to this issue in branch 3.12, so it will be available in the next release.

Thank you for your feedback again.
Best regards.

**Comment by Andy13000ft** [comment]
Thank you for the update.

Regards,
Andy
